File: Leetcode/General/234. Palindrome Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if(head==NULL || head->next==NULL)
            return true;
        ListNode *sp=head;
        ListNode *fp=head;
        ListNode *sp_prev=NULL;
        ////FINDING THE MID oF LINKED LIST
        while(fp!=NULL && fp->next!=NULL)
        {
            sp_prev=sp;
            sp=sp->next;
            fp=fp->next->next;
        }
        
        if(fp!=NULL)                ////[1,2,5,2,1] here fp=1 and [1,2,2,1]  here fp=NULL
        {
            sp=sp->next;
        }
        sp_prev->next=NULL;
        
        ////REVERSING THE HALF oF LINKED LIST
        
        ListNode *cur=sp;
        ListNode *nex=NULL;
        ListNode *prev=NULL;
        while(cur)
        {
            nex=cur->next;
            cur->next=prev;
            prev=cur;
            cur=nex;
        }
        cur=prev;
        
        ListNode* h1=head;
        while(cur && h1)
        {
            if(cur->val==h1->val)
            {
                cur=cur->next;
                h1=h1->next;
            }
            else 
            {
                return false;
            }
        }
        if(cur==NULL && h1==NULL)
            return true;
        return false;
        
    }
};

File: Leetcode/Problems Java/234. Palindrome Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode midOfLL(ListNode head){
        if(head==null || head.next == null)return head;
        ListNode slow = head;
        ListNode fast = head;
        ListNode prevSlow = null;
        while(fast!=null && fast.next!=null){
            prevSlow = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        return prevSlow;
    }
    private ListNode reverse(ListNode head){
        ListNode next = null;
        ListNode curr = head;
        ListNode prev = null;
        while(curr!=null){
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
    private boolean compareLl(ListNode ll1, ListNode ll2){
        while(ll1!=null && ll2!=null){
            if(ll1.val != ll2.val)return false;
            ll1 = ll1.next;
            ll2 = ll2.next;
        }
        return true;
    }
    public boolean isPalindrome(ListNode head) {
        ListNode mid  = midOfLL(head);
        ListNode ll1 = head;
        ListNode ll2 = mid.next;
        mid.next = null;
        
        ll2 = reverse(ll2);
        
        boolean result = compareLl(ll1, ll2);
        
        ll2 = reverse(ll2);
        mid.next = ll2;
        
        return result;
    }
}
